src/v/cloud_topics/level_zero/gc/level_zero_gc.cc

```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
       : health_monitor_(health_monitor)
       , topic_table_(topic_table) {}
 
+    seastar::future<std::expected<partitions_snapshot, std::string>>
+    get_partitions(seastar::abort_source* as) override {
+        const auto& topic_table = topic_table_->local();
+
+        // this revision is for detecting concurrent modifications
+        const auto iter_start_rev = topic_table.topics_map_revision();
+
+        partitions_snapshot snap;
+        snap.last_applied = topic_table.last_applied_revision();
+
+        for (const auto& topic : topic_table.topics_map()) {
+            // we only care about cloud topics
+            if (!topic.second.get_metadata()
+                   .get_configuration()
+                   .is_cloud_topic()) {
+                continue;
+            }
+
+            auto& partitions = snap.partitions[topic.first];
+            for (const auto& partition : topic.second.partitions) {
+                partitions.push_back(model::partition_id(partition.first));
+            }
+
+            co_await seastar::maybe_yield();
+
+            if (as && as->abort_requested()) {
+                co_return std::unexpected("Abort requested");
+            }
+
+            // Detect concurrent changes to the topic table.
+            topic_table.check_topics_map_stable(iter_start_rev);
```

Comment:
I generally try to avoid mixing exceptions and result types. This will throw if there's a concurrent update -- can you catch it and return an unexpected result instead?

Reply:
fixed